File: test/unit/value-text.test.ts
import { describe, expect, it } from 'vitest';
import { LogStore } from '../../src/core/store.ts';
import type { ValueNode } from '../../src/core/types.ts';
import { formatEntryData, formatJson, valueToJson } from '../../src/core/value/data.ts';
import { formatEntryText, formatValueText } from '../../src/core/value/text.ts';
import { createConsole } from '../../src/sources/console/hook.ts';
import { snapshotValue } from '../../src/sources/console/snapshot.ts';

const text = (value: unknown): string => formatValueText(snapshotValue(value), { multiline: true });
const line = (value: unknown): string => formatValueText(snapshotValue(value));
const data = (value: unknown): string => formatJson(valueToJson(snapshotValue(value)));

describe('formatValueText', () => {
	it('keeps every value on one line unless multiline is on', () => {
		const user = { id: 1, roles: ['admin', 'editor'], profile: { city: 'Seoul', zip: '04524' } };

		expect(line(user)).toBe(
			"{ id: 1, roles: ['admin', 'editor'], profile: { city: 'Seoul', zip: '04524' } }"
		);
		expect(text(user)).toBe(
			[
				'{',
				'  id: 1,',
				"  roles: ['admin', 'editor'],",
				"  profile: { city: 'Seoul', zip: '04524' }",
				'}'
			].join('\n')
		);
	});

	it('writes a short object on one line and a long one over several', () => {
		expect(text({ id: 1, name: 'Ada' })).toBe("{ id: 1, name: 'Ada' }");
		expect(
			text({
				id: 1,
				name: 'Ada',
				roles: ['admin', 'editor'],
				profile: { city: 'Seoul', zip: '04524', verified: true }
			})
		).toBe(
			[
				'{',
				'  id: 1,',
				"  name: 'Ada',",
				"  roles: ['admin', 'editor'],",
				"  profile: { city: 'Seoul', zip: '04524', verified: true }",
				'}'
			].join('\n')
		);
	});

	it('writes class names, maps, sets, typed arrays and empty containers', () => {
		class Account {
			id = 7;
		}

		expect(text(new Account())).toBe('Account { id: 7 }');
		expect(
			text(
				new Map<string, unknown>([
					['status', 200],
					['headers', { 'content-type': 'application/json' }]
				])
			)
		).toBe(
			[
				'Map(2) {',
				"  'status' => 200,",
				"  'headers' => { 'content-type': 'application/json' }",
				'}'
			].join('\n')
		);
		expect(text(new Set(['x', 'y']))).toBe("Set(2) { 'x', 'y' }");
		expect(text(new Uint8Array([1, 2]))).toBe('Uint8Array(2) [1, 2]');
		expect(text({})).toBe('{}');
		expect(text([])).toBe('[]');
	});

	it('writes leaves the way previews do, with strings in full', () => {
		const long = 'x'.repeat(60);

		expect(text([undefined, null, -0, 10n, Symbol('token'), /ab+c/gi])).toBe(
			'[undefined, null, -0, 10n, Symbol(token), /ab+c/gi]'
		);
		expect(text({ long })).toBe(`{ long: '${long}' }`);
		expect(text({ quote: "it's\nhere" })).toBe("{ quote: 'it\\'s\\nhere' }");
		// Symbol keys come after string keys, in the order `Reflect.ownKeys` returns them.
		expect(text({ [Symbol('id')]: 1, 'with space': true })).toBe(
			"{ 'with space': true, [Symbol(id)]: 1 }"
		);
	});

	it('marks what the snapshot left out', () => {
		const node: ValueNode = snapshotValue({ deep: { deeper: { list: [1, 2] } } }, { maxDepth: 2 });
		const circular: Record<string, unknown> = {};

		circular.self = circular;

		expect(formatValueText(node)).toBe('{ deep: { deeper: {…} } }');
		expect(text(circular)).toBe('{ self: [Circular] }');
		expect(formatValueText(snapshotValue([1, 2, 3], { maxProperties: 2 }))).toBe(
			'[1, 2, … 1 more]'
		);
		expect(
			formatValueText(
				snapshotValue({
					get value() {
						return 1;
					}
				})
			)
		).toBe('{ value: [Getter] }');
	});

	it('writes an error with its stack trace and its properties', () => {
		const node: ValueNode = {
			kind: 'error',
			className: 'Error',
			value: 'Failed to load',
			stack: '    at load (app.js:1:1)\n    at run (app.js:2:1)',
			children: [
				{
					key: 'cause',
					keyKind: 'property',
					value: { kind: 'error', className: 'TypeError', value: 'Expected a string' }
				}
			]
		};

		expect(formatValueText(node, { multiline: true })).toBe(
			[
				'Error: Failed to load',
				'    at load (app.js:1:1)',
				'    at run (app.js:2:1) { cause: TypeError: Expected a string }'
			].join('\n')
		);
		expect(
			formatValueText(
				{ kind: 'object', children: [{ key: 'error', value: node }] },
				{ multiline: true }
			)
		).toBe(
			[
				'{',
				'  error: Error: Failed to load',
				'      at load (app.js:1:1)',
				'      at run (app.js:2:1) { cause: TypeError: Expected a string }',
				'}'
			].join('\n')
		);
	});

	it('writes an element as markup', () => {
		const link: ValueNode = {
			kind: 'element',
			value: 'a',
			attributes: [['href', '/']],
			children: [{ value: { kind: 'text', value: 'Home' } }]
		};
		const nav: ValueNode = {
			kind: 'element',
			value: 'nav',
			attributes: [['class', 'menu']],
			children: [{ value: link }, { value: { ...link, children: undefined } }]
		};

		expect(formatValueText(link)).toBe('<a href="/">Home</a>');
		expect(
			formatValueText({ kind: 'element', value: 'input', attributes: [['type', 'file']] })
		).toBe('<input type="file">');
		expect(formatValueText(nav)).toBe(
			'<nav class="menu"><a href="/">Home</a><a href="/">…</a></nav>'
		);
		expect(formatValueText(nav, { multiline: true })).toBe(
			['<nav class="menu">', '  <a href="/">Home</a>', '  <a href="/">…</a>', '</nav>'].join('\n')
		);
	});
});

describe('valueToJson and formatJson', () => {
	it('writes objects, arrays, maps and sets as JSON', () => {
		expect(data({ id: 1, name: 'lognal', roles: ['admin', 'editor'] })).toBe(
			'{ "id": 1, "name": "lognal", "roles": ["admin", "editor"] }'
		);
		expect(
			data({
				id: 1,
				name: 'lognal',
				roles: ['admin', 'editor'],
				profile: { city: 'Seoul', zip: '04524', verified: true }
			})
		).toBe(
			[
				'{',
				'  "id": 1,',
				'  "name": "lognal",',
				'  "roles": ["admin", "editor"],',
				'  "profile": { "city": "Seoul", "zip": "04524", "verified": true }',
				'}'
			].join('\n')
		);
		expect(data(new Map([['a', 1]]))).toBe('{ "a": 1 }');
		expect(data(new Map([[1, 'one']]))).toBe('[[1, "one"]]');
		expect(data(new Set(['x', 'y']))).toBe('["x", "y"]');
	});

	it('writes values JSON has no type for as text', () => {
		const circular: Record<string, unknown> = {};

		circular.self = circular;

		expect(
			JSON.parse(data([undefined, null, NaN, -0, 10n, Symbol('token'), /ab+c/gi, new Date(0)]))
		).toEqual([
			null,
			null,
			'NaN',
			0,
			'10n',
			'Symbol(token)',
			'/ab+c/gi',
			'1970-01-01T00:00:00.000Z'
		]);
		expect(data(circular)).toBe('{ "self": "[Circular]" }');
		expect(formatJson(valueToJson(snapshotValue([1, 2, 3], { maxProperties: 2 })))).toBe(
			'[1, 2, "… 1 more"]'
		);
		expect(JSON.parse(data(JSON.parse('{"__proto__": {"polluted": true}}')))).toEqual(
			JSON.parse('{"__proto__": {"polluted": true}}')
		);
		expect(({} as Record<string, unknown>).polluted).toBeUndefined();
	});

	it('writes an error as an object with its name, message and stack', () => {
		const node: ValueNode = {
			kind: 'error',
			className: 'TypeError',
			value: 'Expected a string',
			stack: '    at load (app.js:1:1)',
			children: [{ key: 'code', keyKind: 'property', value: { kind: 'number', value: '42' } }]
		};

		expect(JSON.parse(formatJson(valueToJson(node)))).toEqual({
			name: 'TypeError',
			message: 'Expected a string',
			stack: 'at load (app.js:1:1)',
			code: 42
		});
	});
});

describe('formatEntryText', () => {
	it('joins the text and the values of a console call', () => {
		const store = new LogStore();
		const log = createConsole(store);

		log.log('user', { id: 1 }, 'signed in');

		expect(formatEntryText(store.at(0)!)).toBe('user { id: 1 } signed in');
		expect(formatEntryData(store.at(0)!)).toBe('{ "id": 1 }');

		log.log('plain text');
		expect(formatEntryData(store.at(1)!)).toBe('"plain text"');

		log.log({ a: 1 }, [2]);
		expect(formatEntryData(store.at(2)!)).toBe('[{ "a": 1 }, [2]]');
	});
});

File: packages/flutter/test/core/value_text_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:lognal/src/core/store.dart';
import 'package:lognal/src/core/types.dart';
import 'package:lognal/src/core/value/data.dart';
import 'package:lognal/src/core/value/text.dart';
import 'package:lognal/src/sources/console/capture.dart';
import 'package:lognal/src/sources/console/console.dart';

String multiline(Object? value) {
  return formatValueText(captureValue(value), const ValueTextOptions(multiline: true));
}

String line(Object? value) => formatValueText(captureValue(value));

String data(Object? value) => formatJson(valueToJson(captureValue(value)));

class Account {
  const Account();

  Map<String, Object?> toJson() => <String, Object?>{'id': 7};
}

void main() {
  group('formatValueText', () {
    test('keeps every value on one line unless multiline is on', () {
      const Map<String, Object?> user = <String, Object?>{
        'id': 1,
        'roles': <String>['admin', 'editor'],
        'profile': <String, String>{'city': 'Seoul', 'zip': '04524'},
      };

      expect(
        line(user),
        "Map(3) { 'id': 1, 'roles': ['admin', 'editor'], "
        "'profile': Map(2) { 'city': 'Seoul', 'zip': '04524' } }",
      );
      expect(
        multiline(user),
        <String>[
          'Map(3) {',
          "  'id': 1,",
          "  'roles': ['admin', 'editor'],",
          "  'profile': Map(2) { 'city': 'Seoul', 'zip': '04524' }",
          '}',
        ].join('\n'),
      );
    });

    test('writes a short value on one line and a long one over several', () {
      expect(
        multiline(<String, Object?>{'id': 1, 'name': 'Ada'}),
        "Map(2) { 'id': 1, 'name': 'Ada' }",
      );
      expect(
        multiline(<String, Object?>{
          'id': 1,
          'name': 'Ada',
          'roles': <String>['admin', 'editor'],
          'profile': <String, Object?>{'city': 'Seoul', 'zip': '04524', 'verified': true},
        }),
        <String>[
          'Map(4) {',
          "  'id': 1,",
          "  'name': 'Ada',",
          "  'roles': ['admin', 'editor'],",
          "  'profile': Map(3) { 'city': 'Seoul', 'zip': '04524', 'verified': true }",
          '}',
        ].join('\n'),
      );
    });

    test('writes type names, sets and empty containers', () {
      expect(multiline(const Account()), 'Account { id: 7 }');
      expect(multiline(<String>{'x', 'y'}), "Set(2) { 'x', 'y' }");
      expect(multiline(<String, Object?>{}), 'Map(0) {}');
      expect(multiline(<Object?>[]), '[]');
    });

    test('writes leaves the way previews do, with strings in full', () {
      final String long = 'x' * 40;

      expect(
        multiline(<Object?>[null, -0.0, BigInt.from(10), RegExp('ab+c')]),
        '[null, -0.0, 10, /ab+c/]',
      );
      expect(multiline(<String, Object?>{'long': long}), "Map(1) { 'long': '$long' }");
      expect(
        multiline(<String, Object?>{'quote': "it's\nhere"}),
        "Map(1) { 'quote': 'it\\'s\\nhere' }",
      );
    });

    test('marks what the capture left out', () {
      const Map<String, Object?> deep = <String, Object?>{
        'deep': <String, Object?>{
          'deeper': <String, Object?>{
            'list': <int>[1, 2],
          },
        },
      };
      final Map<String, Object?> circular = <String, Object?>{};

      circular['self'] = circular;

      expect(
        formatValueText(captureValue(deep, const CaptureOptions(maxDepth: 2))),
        "Map(1) { 'deep': Map(1) { 'deeper': Map(1) {…} } }",
      );
      expect(multiline(circular), "Map(1) { 'self': [Circular] }");
      expect(
        formatValueText(captureValue(<int>[1, 2, 3], const CaptureOptions(maxProperties: 2))),
        '[1, 2, … 1 more]',
      );
    });

    test('writes an error with its stack trace and its properties', () {
      const ValueNode node = ValueNode(
        kind: ValueKind.error,
        className: 'StateError',
        value: 'Failed to load',
        stack: '    at load (app.dart:1:1)\n    at run (app.dart:2:1)',
        children: <ValueEntry>[
          ValueEntry(
            key: 'cause',
            keyKind: ValueKeyKind.property,
            value: ValueNode(
              kind: ValueKind.error,
              className: 'FormatException',
              value: 'Expected a string',
            ),
          ),
        ],
      );

      expect(
        formatValueText(node, const ValueTextOptions(multiline: true)),
        <String>[
          'StateError: Failed to load',
          '    at load (app.dart:1:1)',
          '    at run (app.dart:2:1) { cause: FormatException: Expected a string }',
        ].join('\n'),
      );
    });

    test('writes an element as markup', () {
      const ValueNode link = ValueNode(
        kind: ValueKind.element,
        value: 'a',
        attributes: <MapEntry<String, String>>[MapEntry<String, String>('href', '/')],
        children: <ValueEntry>[
          ValueEntry(
            value: ValueNode(kind: ValueKind.text, value: 'Home'),
          ),
        ],
      );
      const ValueNode nav = ValueNode(
        kind: ValueKind.element,
        value: 'nav',
        attributes: <MapEntry<String, String>>[MapEntry<String, String>('class', 'menu')],
        children: <ValueEntry>[
          ValueEntry(value: link),
          ValueEntry(
            value: ValueNode(
              kind: ValueKind.element,
              value: 'a',
              attributes: <MapEntry<String, String>>[MapEntry<String, String>('href', '/')],
            ),
          ),
        ],
      );

      expect(formatValueText(link), '<a href="/">Home</a>');
      expect(
        formatValueText(
          const ValueNode(
            kind: ValueKind.element,
            value: 'input',
            attributes: <MapEntry<String, String>>[MapEntry<String, String>('type', 'file')],
          ),
        ),
        '<input type="file">',
      );
      expect(formatValueText(nav), '<nav class="menu"><a href="/">Home</a><a href="/">…</a></nav>');
      expect(
        formatValueText(nav, const ValueTextOptions(multiline: true)),
        <String>[
          '<nav class="menu">',
          '  <a href="/">Home</a>',
          '  <a href="/">…</a>',
          '</nav>',
        ].join('\n'),
      );
    });
  });

  group('valueToJson and formatJson', () {
    test('writes maps, lists and sets as JSON', () {
      expect(
        data(<String, Object?>{
          'id': 1,
          'name': 'lognal',
          'roles': <String>['admin', 'editor'],
        }),
        '{ "id": 1, "name": "lognal", "roles": ["admin", "editor"] }',
      );
      expect(
        data(<String, Object?>{
          'id': 1,
          'name': 'lognal',
          'roles': <String>['admin', 'editor'],
          'profile': <String, Object?>{'city': 'Seoul', 'zip': '04524', 'verified': true},
        }),
        <String>[
          '{',
          '  "id": 1,',
          '  "name": "lognal",',
          '  "roles": ["admin", "editor"],',
          '  "profile": { "city": "Seoul", "zip": "04524", "verified": true }',
          '}',
        ].join('\n'),
      );
      expect(data(<int, String>{1: 'one'}), '[[1, "one"]]');
      expect(data(<String>{'x', 'y'}), '["x", "y"]');
    });

    test('writes values JSON has no type for as text', () {
      final Map<String, Object?> circular = <String, Object?>{};

      circular['self'] = circular;

      expect(
        jsonDecode(
          data(<Object?>[null, double.nan, BigInt.from(10), RegExp('ab+c'), DateTime.utc(1970)]),
        ),
        <Object?>[null, 'NaN', '10', '/ab+c/', '1970-01-01T00:00:00.000Z'],
      );
      expect(data(circular), '{ "self": "[Circular]" }');
      expect(
        formatJson(
          valueToJson(captureValue(<int>[1, 2, 3], const CaptureOptions(maxProperties: 2))),
        ),
        '[1, 2, "… 1 more"]',
      );
    });

    test('writes an error as an object with its name, message and stack', () {
      const ValueNode node = ValueNode(
        kind: ValueKind.error,
        className: 'FormatException',
        value: 'Expected a string',
        stack: '    at load (app.dart:1:1)',
        children: <ValueEntry>[
          ValueEntry(
            key: 'code',
            keyKind: ValueKeyKind.property,
            value: ValueNode(kind: ValueKind.number, value: '42'),
          ),
        ],
      );

      expect(jsonDecode(formatJson(valueToJson(node))), <String, Object?>{
        'name': 'FormatException',
        'message': 'Expected a string',
        'stack': 'at load (app.dart:1:1)',
        'code': 42,
      });
    });
  });

  group('formatEntryText', () {
    test('joins the text and the values of a logging call', () {
      final LogStore store = LogStore();
      final LognalConsole log = LognalConsole(store);

      log.log('user', <Object?>[
        <String, int>{'id': 1},
        'signed in',
      ]);

      expect(formatEntryText(store.at(0)!), "user Map(1) { 'id': 1 } signed in");
      expect(formatEntryData(store.at(0)!), '{ "id": 1 }');

      log.log('plain text');
      expect(formatEntryData(store.at(1)!), '"plain text"');

      log.log(
        <String, int>{'a': 1},
        <Object?>[
          <int>[2],
        ],
      );
      expect(formatEntryData(store.at(2)!), '[{ "a": 1 }, [2]]');
    });
  });
}
